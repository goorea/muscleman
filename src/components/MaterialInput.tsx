import React from 'react';
import { TextField, TextFieldProps } from 'rn-material-ui-textfield';
import { Icon } from 'react-native-elements';
import {
  Animated,
  NativeSyntheticEvent,
  TextInputChangeEventData,
  TextInputFocusEventData,
} from 'react-native';
import { useTheme } from '@src/contexts/ThemeProvider';

type P = TextFieldProps & {
  onClear?: () => void;
};

const MaterialInput: React.ForwardRefRenderFunction<TextField, P> = (
  props,
  ref,
) => {
  const { colors } = useTheme();
  const fadeAnimation = new Animated.Value(0);
  const fade = (focus: boolean) =>
    Animated.timing(fadeAnimation, {
      toValue: focus ? 1 : 0,
      duration: 225,
      useNativeDriver: true,
    }).start();
  const onFocus = (event: NativeSyntheticEvent<TextInputFocusEventData>) => {
    fade(true);

    if (props.onFocus) {
      props.onFocus(event);
    }
  };
  const onBlur = (event: NativeSyntheticEvent<TextInputChangeEventData>) => {
    fade(false);

    if (props.onBlur) {
      props.onBlur(event);
    }
  };
  const renderRightAccessory = () => (
    <Animated.View style={{ opacity: fadeAnimation }}>
      <Icon
        testID="icon"
        onPress={props.onClear}
        name="close-circle"
        type="ionicon"
        color={colors.grey4}
      />
    </Animated.View>
  );

  return (
    <TextField
      ref={ref}
      textColor={colors.black}
      labelFontSize={14}
      lineWidth={1}
      activeLineWidth={1}
      disabledLineWidth={1}
      tintColor={colors.grey3}
      baseColor={colors.grey3}
      errorColor={colors.error}
      disabledLineType="solid"
      renderRightAccessory={renderRightAccessory}
      {...props}
      onFocus={onFocus}
      onBlur={onBlur}
    />
  );
};

export default React.forwardRef(MaterialInput);
