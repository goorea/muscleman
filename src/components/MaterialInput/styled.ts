import styled from 'styled-components/native';
import { Animated } from 'react-native';
import { Colors } from '@src/types/theme';
import { fonts } from '@src/theme';

export const Container = styled.View``;

export const Wrapper = styled.View`
  margin-bottom: 4px;
  padding: 10px 0;
  border-bottom-width: 1px;
  border-bottom-color: ${({ theme }) => theme.greyOutline};
`;

export const Label = styled(Animated.Text)<{ color: keyof Colors }>`
  font-family: ${fonts.normal};
  font-size: 16px;
  color: ${({ theme, color }) => theme[color]};
`;

export const Input = styled.TextInput`
  font-family: ${fonts.normal};
  font-size: 16px;
  color: ${({ theme }) => theme.foreground};
  padding: 0;
`;
