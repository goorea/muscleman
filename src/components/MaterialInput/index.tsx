import React, { useState } from 'react';
import {
  LayoutChangeEvent,
  LayoutRectangle,
  NativeSyntheticEvent,
  TextInput,
  TextInputFocusEventData,
  TextInputProps,
} from 'react-native';
import Text from '@src/components/Text';
import useAnimation from './hooks/useAnimation';
import { Container, Label, Wrapper, Input } from './styled';

type P = TextInputProps & {
  onChange: (text: string) => void;
  label: string;
  error?: string;
};

const MaterialInput: React.ForwardRefRenderFunction<TextInput, P> = (
  props,
  ref,
) => {
  const [active, setActive] = useState<boolean>(false);
  const [labelLayout, setLabelLayout] = useState<
    Pick<LayoutRectangle, 'width' | 'height'>
  >({
    width: 0,
    height: 0,
  });
  const onLabelLayout = ({ nativeEvent }: LayoutChangeEvent) => {
    setLabelLayout(nativeEvent.layout);
  };
  const onFocus = (event: NativeSyntheticEvent<TextInputFocusEventData>) => {
    if (props.onFocus) {
      props.onFocus(event);
    }

    setActive(true);
  };
  const onBlur = (event: NativeSyntheticEvent<TextInputFocusEventData>) => {
    if (props.onBlur) {
      props.onBlur(event);
    }

    if (!props.value) {
      setActive(false);
    }
  };

  const { translateX, translateY, scale } = useAnimation(active, labelLayout);

  return (
    <Container>
      <Wrapper>
        <Label
          onLayout={onLabelLayout}
          color={active ? 'grey3' : 'foreground'}
          style={{ transform: [{ translateX }, { translateY }, { scale }] }}>
          {props.label}
        </Label>
        <Input
          ref={ref}
          {...props}
          onChangeText={props.onChange}
          onFocus={onFocus}
          onBlur={onBlur}
        />
      </Wrapper>
      {!!props.error && (
        <Text size={12} color="error">
          {props.error}
        </Text>
      )}
    </Container>
  );
};

export default React.forwardRef(MaterialInput);
