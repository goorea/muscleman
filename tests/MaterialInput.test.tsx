import 'react-native';
import React from 'react';
import MaterialInput from '@src/components/MaterialInput';
import { render } from '@testing-library/react-native';

describe('MaterialInput 컴포넌트', () => {
  it('렌더링이 올바르게 된다', () => {
    const { toJSON } = render(<MaterialInput />);

    expect(toJSON()).toMatchSnapshot();
  });
});
