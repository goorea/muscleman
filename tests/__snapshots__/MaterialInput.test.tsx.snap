// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MaterialInput 컴포넌트 렌더링이 올바르게 된다 1`] = `
<View
  style={
    Array [
      Object {},
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "baseline",
          "borderBottomColor": "#bbb",
          "borderBottomWidth": 1,
          "marginBottom": 4,
          "paddingBottom": 10,
          "paddingLeft": 0,
          "paddingRight": 0,
          "paddingTop": 10,
        },
      ]
    }
  >
    <Text
      collapsable={false}
      color="foreground"
      nativeID="animatedComponent"
      onLayout={[Function]}
      style={
        Object {
          "color": "#000000",
          "fontFamily": "SpoqaHanSansNeo-Regular",
          "fontSize": 16,
          "transform": Array [
            Object {
              "translateX": 0,
            },
            Object {
              "translateY": 16,
            },
            Object {
              "scale": 1,
            },
          ],
        }
      }
    >
      이름
    </Text>
    <TextInput
      label="이름"
      onBlur={[Function]}
      onChange={[MockFunction]}
      onChangeText={[MockFunction]}
      onFocus={[Function]}
      style={
        Array [
          Object {
            "color": "#000000",
            "fontFamily": "SpoqaHanSansNeo-Regular",
            "fontSize": 16,
            "paddingBottom": 0,
            "paddingLeft": 0,
            "paddingRight": 0,
            "paddingTop": 0,
            "width": "100%",
          },
        ]
      }
    />
  </View>
</View>
`;
